# **Evaluación corta — Regresión múltiple con datos de próstata**

El objetivo es estudiar variables asociadas al logaritmo del antígeno prostático específico (lpsa).



In [ ]:
suppressPackageStartupMessages(
  library(dplyr)
)

In [ ]:
#Cargamos la base de datos
df_prostate = read.table("/content/prostate.txt",header = TRUE)
#Visualizamos las primeras 6 filas
head(df_prostate)

,lcavol,lweight,age,lbph,svi,lcp,gleason,pgg45,lpsa,train
,<dbl>,<dbl>,<int>,<dbl>,<int>,<dbl>,<int>,<int>,<dbl>,<lgl>
1,-0.5798185,2.769459,50,-1.386294,0,-1.386294,6,0,-0.4307829,TRUE
2,-0.9942523,3.319626,58,-1.386294,0,-1.386294,6,0,-0.1625189,TRUE
3,-0.5108256,2.691243,74,-1.386294,0,-1.386294,7,20,-0.1625189,TRUE
4,-1.2039728,3.282789,58,-1.386294,0,-1.386294,6,0,-0.1625189,TRUE
5,0.7514161,3.432373,62,-1.386294,0,-1.386294,6,0,0.3715636,TRUE
6,-1.0498221,3.228826,50,-1.386294,0,-1.386294,6,0,0.7654678,TRUE


 1. (1 punto) **Inspección de la base de datos.** La función str nos permite ver la estructura de la base de datos.

*   **Pregunta:** A partir de str(df_prostate), indique cuántas observaciones y variables tiene la base.
*   **Respuesta:** La base tiene 97 observaciones y 10 variables.


In [ ]:
#La fun
str(df_prostate)

'data.frame':	97 obs. of  10 variables:
 $ lcavol : num  -0.58 -0.994 -0.511 -1.204 0.751 ...
 $ lweight: num  2.77 3.32 2.69 3.28 3.43 ...
 $ age    : int  50 58 74 58 62 50 64 58 47 63 ...
 $ lbph   : num  -1.39 -1.39 -1.39 -1.39 -1.39 ...
 $ svi    : int  0 0 0 0 0 0 0 0 0 0 ...
 $ lcp    : num  -1.39 -1.39 -1.39 -1.39 -1.39 ...
 $ gleason: int  6 6 7 6 6 6 6 6 6 6 ...
 $ pgg45  : int  0 0 20 0 0 0 0 0 0 0 ...
 $ lpsa   : num  -0.431 -0.163 -0.163 -0.163 0.372 ...
 $ train  : logi  TRUE TRUE TRUE TRUE TRUE TRUE ...


2. (1 punto) **Intepretación de código.**

*   **Pregunta:** Explique de manera breve, en sus propias palabras qué esta haciendo el siguiente código.
*   **Respuesta:**
La celda está dividiendo la base en un set de entrenamiento y de testeo, usando como filtro la columna "train".


In [ ]:

df_train <- df_prostate |> filter(train==TRUE)
print(dim(df_train))

df_test <- df_prostate |> filter(train==FALSE)
print(dim(df_test))

[1] 67 10
[1] 30 10


3. (1 punto) **Construcción del modelo** Considere inicialmente las variables `lcavol, lweight, age, svi`.


*   Ajuste un modelo lineal para explicar lpsa a partir de estas cuatro variables. Utilice `df_train`.
*   Escriba la ecuación del modelo estimado usando **los coeficientes que obtuvo en R.**
* De las cuatro variables, indique cuál presenta mayor evidencia de asociación con lpsa.

Incluya código utilizado y el resultado numérico que justifíca su respuesta.

In [ ]:
# construcción del modelo lineal
modelo_lpsa <- lm(lpsa ~ lcavol + lweight + age + svi, data = df_train)
modelo_lpsa


Call:
lm(formula = lpsa ~ lcavol + lweight + age + svi, data = df_train)

Coefficients:
(Intercept)       lcavol      lweight          age          svi  
   -0.64219      0.53173      0.76812     -0.00784      0.52891  


In [ ]:
summary(modelo_lpsa)


Call:
lm(formula = lpsa ~ lcavol + lweight + age + svi, data = df_train)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.69266 -0.48898 -0.00507  0.49826  1.69232 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -0.64219    0.96342  -0.667  0.50752    
lcavol       0.53173    0.09697   5.484 8.13e-07 ***
lweight      0.76812    0.20941   3.668  0.00051 ***
age         -0.00784    0.01326  -0.591  0.55660    
svi          0.52891    0.27277   1.939  0.05705 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.7482 on 62 degrees of freedom
Multiple R-squared:  0.6395,	Adjusted R-squared:  0.6162 
F-statistic: 27.49 on 4 and 62 DF,  p-value: 3.834e-13


**ecuación del modelo estimado con coeficientes obtenidos**

lpsa = -0.64219 + 0.76812\*lweight + 0.53173\*lcavol + 0.52891\*svi - 0.00784\*age

La variable con mayor evidencia de asociación con lpsa es "lcavol" (log-volumen del cáncer): su p-value es 8.13e-07, el más pequeño de las cuatro variables (t = 5.48).




4. (1 punto) **No basta con mirar el p-valor**:
Seleccione una variable (la que usted quiera) que sea significativa a nivel α=0.1 en su modelo. Explique en **dos lineas** qué significa su coeficiente estimado en el contexto del problema. Luego responda: ¿sería correcto decir que esta variable causa un aumento del PSA? Justifique su respuesta.

el nivel de significancia indica qué variables son estadísticamente relevantes para el modelo. cuando p value < α, entonces se rechaza la hipotesis nula (H0 : coef = 0), y nuestra variable es relevante para explicar la función objetivo.

en el caso de "lcavol" su coeficiente estimado (0.532) indica que, ante un aumento de una unidad en el log-volumen del tumor, lpsa aumenta en promedio 0.532 unidades, manteniendo constantes las demás variables del modelo.

no es del todo correcto decir que, en este caso, "lcavol" causa un aumento del psa, en este caso solo se muestra que existe una relación significativa, sin explicitar si existe una causalidad entre ambas.

In [ ]:
summary(modelo_lpsa)


Call:
lm(formula = lpsa ~ lcavol + lweight + age + svi, data = df_train)

Residuals:
     Min       1Q   Median       3Q      Max 
-1.69266 -0.48898 -0.00507  0.49826  1.69232 

Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept) -0.64219    0.96342  -0.667  0.50752    
lcavol       0.53173    0.09697   5.484 8.13e-07 ***
lweight      0.76812    0.20941   3.668  0.00051 ***
age         -0.00784    0.01326  -0.591  0.55660    
svi          0.52891    0.27277   1.939  0.05705 .  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

Residual standard error: 0.7482 on 62 degrees of freedom
Multiple R-squared:  0.6395,	Adjusted R-squared:  0.6162 
F-statistic: 27.49 on 4 and 62 DF,  p-value: 3.834e-13


5. (2 puntos) **¿Conviene complicar el modelo?** Ajuste ahora:

In [ ]:
modelo_2 <- lm(lpsa ~ lcavol + lweight + age + svi +lbph + lcp + gleason + pgg45,
               data = df_train)

Compare este modelo con el modelo anterior usadno:
* R^2 ajustado
* AIC
Si tuviera que escoger solo uno de los dos modelos ¿Cuál escogería?

In [ ]:
summary(modelo_lpsa)$adj.r.squared # para ver los r^2 ajustados
summary(modelo_2)$adj.r.squared
AIC(modelo_lpsa, modelo_2)

[1] 0.6162124

[1] 0.6522155

,df,AIC
,<dbl>,<dbl>
modelo_lpsa,6,158.0783
modelo_2,10,155.0101


considerando los coeficientes obtenidos, el modelo 2 supera al modelo 1 (modelo_lpsa) al tener un mayor r^2 ajustado y un menor AIC, por lo que es el modelo elegido.

usamos el r^2 ajustado y no el normal ya que el primero permite evidenciar la penalización de agregar nuevas variables que empeoran el modelo.

Considere el siguiente código, que obtiene el error predictivo asociado al segundo modelo.

In [ ]:
pred_2 <- predict(modelo_2, newdata = df_test)
mean((df_test$lpsa - pred_2)^2)

[1] 0.521274

Replique el código anterior para obtener el error predictivo asociado al primer modelo que ajusto. ¿El modelo que prefirió usando AIC y R^2 ajustado es también el que presenta menor error de predicción en df_test? Comente brevemente.

In [ ]:
pred_lpsa <- predict(modelo_lpsa, newdata = df_test)
mean((df_test$lpsa - pred_lpsa)^2)

[1] 0.3878768

no, el modelo preferido usando el r^2 ajustado y el AIC no es el mismo que tiene el menor error de predicción df_test. esto ocurre porque ambos indicadores se calculan usando los mismos datos de estimación, por lo que no garantizan un buen resultado con nuevos datos.

